homeassistant/components/music_assistant/media_player.py

```diff
@@ -0,0 +1,751 @@
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    ATTR_ACTIVE_GROUP,
+    ATTR_ACTIVE_QUEUE,
+    ATTR_GROUP_LEADER,
+    ATTR_MASS_PLAYER_ID,
+    ATTR_MASS_PLAYER_TYPE,
+    ATTR_QUEUE_INDEX,
+    ATTR_QUEUE_ITEMS,
+    ATTR_STREAM_TITLE,
+    DOMAIN,
+)
+from .entity import MassBaseEntity
+from .helpers import get_mass
+
+if TYPE_CHECKING:
+    from music_assistant.client import MusicAssistantClient
+    from music_assistant.common.models.player import Player
+    from music_assistant.common.models.player_queue import PlayerQueue
+
+SUPPORTED_FEATURES = (
+    MediaPlayerEntityFeature.PAUSE
+    | MediaPlayerEntityFeature.VOLUME_SET
+    | MediaPlayerEntityFeature.STOP
+    | MediaPlayerEntityFeature.PREVIOUS_TRACK
+    | MediaPlayerEntityFeature.NEXT_TRACK
+    | MediaPlayerEntityFeature.SHUFFLE_SET
+    | MediaPlayerEntityFeature.REPEAT_SET
+    | MediaPlayerEntityFeature.TURN_ON
+    | MediaPlayerEntityFeature.TURN_OFF
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.PLAY_MEDIA
+    | MediaPlayerEntityFeature.VOLUME_STEP
+    | MediaPlayerEntityFeature.CLEAR_PLAYLIST
+    | MediaPlayerEntityFeature.BROWSE_MEDIA
+    | MediaPlayerEntityFeature.MEDIA_ENQUEUE
+    | MediaPlayerEntityFeature.MEDIA_ANNOUNCE
+    | MediaPlayerEntityFeature.SEEK
+)
+
+STATE_MAPPING = {
+    PlayerState.IDLE: STATE_IDLE,
+    PlayerState.PLAYING: STATE_PLAYING,
+    PlayerState.PAUSED: STATE_PAUSED,
+}
+
+QUEUE_OPTION_MAP = {
+    # map from HA enqueue options to MA enqueue options
+    # which are the same but just in case
+    MediaPlayerEnqueue.ADD: QueueOption.ADD,
+    MediaPlayerEnqueue.NEXT: QueueOption.NEXT,
+    MediaPlayerEnqueue.PLAY: QueueOption.PLAY,
+    MediaPlayerEnqueue.REPLACE: QueueOption.REPLACE,
+}
+
+ATTR_RADIO_MODE = "radio_mode"
+ATTR_MEDIA_ID = "media_id"
+ATTR_MEDIA_TYPE = "media_type"
+ATTR_ARTIST = "artist"
+ATTR_ALBUM = "album"
+ATTR_URL = "url"
+ATTR_USE_PRE_ANNOUNCE = "use_pre_announce"
+ATTR_ANNOUNCE_VOLUME = "announce_volume"
+ATTR_SOURCE_PLAYER = "source_player"
+ATTR_AUTO_PLAY = "auto_play"
+
+# pylint: disable=too-many-public-methods
+
+_MassPlayerT = TypeVar("_MassPlayerT", bound="MassPlayer")
+_R = TypeVar("_R")
+_P = ParamSpec("_P")
+
+
+def catch_musicassistant_error(
+    func: Callable[Concatenate[_MassPlayerT, _P], Awaitable[_R]],
+) -> Callable[Concatenate[_MassPlayerT, _P], Coroutine[Any, Any, _R | None]]:
+    """Check and log commands to players."""
+
+    @functools.wraps(func)
+    async def wrapper(
+        self: _MassPlayerT, *args: _P.args, **kwargs: _P.kwargs
+    ) -> _R | None:
+        """Catch Music Assistant errors and convert to Home Assistant error."""
+        try:
+            return await func(self, *args, **kwargs)
+        except MusicAssistantError as err:
+            raise HomeAssistantError(f"{err.__class__.__name__}: {err!s}") from err
```

Comment:
Instead of showing the user an error a dev can understand, I would rather see user facing errors like "Could not add to queue" or something along those lines (they can even be translated)

Reply:
Let's do that in a follow up PR.
We attach codes to the MusicAssistant errors so we could use those to translate into translation keys.
IMO that is out of scope of this base PR

Reply:
BTW: The errors are human readable ;-)

Reply:
`err.__class__.__name__` gives the idea that it's the name of the exception class, which is not user friendly in general

Reply:
Sure but those are human readable - again this is something for a follow up PR, not in scope for now.
Lets do follow up PR where we map the MA error codes to translation keys

Reply:
Do I need to make changes here?

Reply:
I would leave this as-is now and then do a small follow-up PR where we map MA errors into HA translation keys - that way the errors will even be translated.

One small adjustment could be to change from this:

```
raise HomeAssistantError(f"{err.__class__.__name__}: {err!s}") from err
```

to this:

```
error_msg = str(err) or err.__class__.__name__
raise HomeAssistantError(error_msg) from err
```

Because in 99% if the cases the exception message will already be an english, human readable string and only in case there is no exception message, use the exception's class name as fallback.

The follow up PR would then map the error_code within the Music Assistant exceptions to a translation key